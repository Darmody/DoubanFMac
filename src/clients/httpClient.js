import R from 'ramda'
import { defaults } from 'utils/ramda'
import qs from 'qs'
import type { Prop } from 'constants/types/Client'

const URLENCODE_TYPE = 'application/x-www-form-urlencoded'
const JSON_TYPE = 'application/json'


const defaultProps = (props: Prop): Prop => defaults({
  method: 'GET',
  endpoint: '',
  queryString: {},
  body: {},
  headers: {},
  json: false,
  credentials: 'same-origin',
})(props)

const normalizeProps = (props: Prop): Prop => ({
  ...props,
  method: props.method.toUpperCase(),
})

const jsonify = (props: Prop): Prop => {
  const jsonifyHeaders = headers => ({
    ...headers,
    'Content-Type': JSON_TYPE,
  })
  const jsonifyBody = body => JSON.stringify(body)

  const jsonifyOr = R.ifElse(props.json)

  return {
    ...props,
    headers: jsonifyOr(jsonifyHeaders, props.headers),
    body: jsonifyOr(jsonifyBody, props.body)
  }
}

const formUrlencode = (props: Prop): Prop => {
  const urlencoded = props.headers['Content-Type'] === URLENCODE_TYPE

  return {
    ...props,
    body: urlencoded ? qs.stringify(props.body) : props.body,
  }
}

const removeBodyWhenGet = (props: Prop): Prop => {
  const isGet = props.method === 'GET'

  return {
    ...props,
    body: isGet ? undefined : props.body,
  }
}

const httpClient = (props: Prop): Prop => {
  const finalProps = R.compose(
    removeBodyWhenGet,
    formUrlencode,
    jsonify,
    normalizeProps,
    defaultProps,
  )(props)

  const { endpoint, queryString, ...fetchParams } = finalProps
  const url = `${endpoint}?${qs.stringify(queryString)}`

  return fetch(url, fetchParams)
}

export default httpClient
