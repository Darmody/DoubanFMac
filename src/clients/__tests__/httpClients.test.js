import httpClient from '../httpClient'

describe('httpClient', () => {
  beforeEach(() => {
    window.fetch = jest.fn()
  })

  test('设置好初始值', () => {
    httpClient({})

    const fetchCalls = window.fetch.mock.calls
    expect(fetchCalls.length).toBe(1)
    expect(fetchCalls[0][1]).toMatchSnapshot('httpclient::defaultValue')

    httpClient()

    expect(window.fetch.mock.calls.length).toMatchSnapshot('httpclient::defaultValue')

  })

  test('method 自动转化为大写', () => {
    httpClient({ method: 'get' })

    const fetchCalls = window.fetch.mock.calls
    expect(fetchCalls.length).toBe(1)
    expect(fetchCalls[0][1].method).toBe('GET')
  })

  test('处理 json 类型的请求', () => {
    const body = { field: 'value' }

    httpClient({
      method: 'post',
      body,
      json: true,
    })

    const fetchCalls = window.fetch.mock.calls
    expect(fetchCalls.length).toBe(1)
    expect(JSON.parse(fetchCalls[0][1].body)).toMatchObject(body)
    expect(fetchCalls[0][1].headers['Content-Type']).toBe('application/json')
  })

  test('处理 form-urlencode 类型的请求', () => {
    const body = { field: 'value' }
    const contentType = 'application/x-www-form-urlencoded'

    httpClient({
      method: 'post',
      body,
      headers: {
        'Content-Type': contentType,
      }
    })

    const fetchCalls = window.fetch.mock.calls
    expect(fetchCalls.length).toBe(1)
    expect(fetchCalls[0][1].body).toBe('field=value')
    expect(fetchCalls[0][1].headers['Content-Type']).toBe(contentType)
  })

  test('当 get 请求时, 去除 body', () => {
    const body = { field: 'value' }

    httpClient({
      method: 'get',
      body,
    })

    const fetchCalls = window.fetch.mock.calls
    expect(fetchCalls.length).toBe(1)
    expect(fetchCalls[0][1].body).toBeUndefined()
  })
})
