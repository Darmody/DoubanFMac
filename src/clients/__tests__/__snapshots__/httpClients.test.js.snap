// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`httpclient::defaultValue 1`] = `
Object {
  "body": undefined,
  "credentials": "same-origin",
  "headers": Object {},
  "json": false,
  "method": "GET",
}
`;

exports[`httpclient::defaultValue 2`] = `2`;
